src/wsproxy.ml

```diff
@@ -15,124 +15,115 @@
     | "hixie76" ->
-      Printf.printf "old-style\n%!";
-      (wsframe_old, wsunframe_old)
+      Lwt_log.debug_f "old-style (hixie76) protocol" >>= fun () ->
+      Lwt.return (wsframe_old, wsunframe_old)
     | "hybi10" ->
-      Printf.printf "new-style\n%!";
-      (wsframe, wsunframe)
+      Lwt_log.debug_f "new-style (hybi10) protocol" >>= fun () ->
+      Lwt.return (wsframe, wsunframe)
     | _ ->
-      Printf.printf "unknown\n%!";
-      (wsframe,wsunframe)
-  in
-  let (realframe,realunframe) = (frame, unframe) in
-  let fdno = Printf.sprintf "%d" (Obj.magic fd) in
-  open_connection_fd "localhost" localport >>= fun localfd ->
-  Lwt.finalize
-    (fun () ->
-       let thread1 =
-         lwt_fd_enumerator localfd (realframe (writer (really_write fd) "thread1")) >>= fun _ ->
-         Lwt.return () in
-       let thread2 =
-         lwt_fd_enumerator fd (realunframe (writer (really_write localfd) "thread2")) >>= fun _ ->
-         Lwt.return () in
-       ignore_exn (fun () -> Lwt.join [thread1; thread2]) ()
-    )
-    (fun _ ->
-       ignore_exn (fun () -> Lwt_unix.close fd) () >>= fun () ->
-       ignore_exn (fun () -> Lwt_unix.close localfd) ())
-  >>= fun () ->
-  Printf.printf "FD %s pipe closed: %b %b\n" fdno Lwt_unix.(state fd == Closed) Lwt_unix.(state localfd == Closed)
-  |> Lwt.return
+      Lwt_log.debug_f "unknown protocol" >>= fun () ->
+      Lwt.return (wsframe, wsunframe) 
+  end >>= fun (frame,unframe) ->
+  with_open_connection_fd "localhost" localport ~callback:(fun localfd ->
+      let session_id = Uuidm.v `V4 |> Uuidm.to_string in
+      Lwt_log.debug_f "Starting proxy session %s" session_id >>= fun () ->
+      let thread1 =
+        lwt_fd_enumerator localfd (frame (writer (really_write fd) "thread1")) >>= fun _ ->
+        Lwt.return_unit in
+      let thread2 =
+        lwt_fd_enumerator fd (unframe (writer (really_write localfd) "thread2")) >>= fun _ ->
+        Lwt.return_unit in
+      (* closing the connection in one of the threads above in general leaves the other pending forever,
+       * by using choose here, we make sure that as soon as one of the threads completes, both are closed *)
+      Lwt.choose [thread1; thread2]
+      >>= fun () -> Lwt_log.debug_f "Closing proxy session %s" session_id)
+
 
 let handler sock msg =
-  Lwt_io.printf "Got msg: %s\n" msg >>= fun _ ->
+  Lwt_log.debug_f "Got msg: %s" msg >>= fun () ->
   match Re_str.split (Re_str.regexp "[:]") msg with
   | [protocol;ty;sport] ->
     let port = int_of_string sport in
     proxy sock protocol ty port
   | [protocol;sport] ->
     let port = int_of_string sport in
     proxy sock protocol "hybi10" port
-  | _ ->
-    let sid = Printf.sprintf "%d" (Obj.magic (Lwt_unix.unix_file_descr sock)) in
-    ignore_exn (fun () -> Lwt_unix.close sock) () >>= fun () ->
-    Printf.printf "Malformed message. Socket %s closed: %b\n" sid Lwt_unix.(state sock == Closed)
-    |> Lwt.return
+  | _ -> Lwt.return_unit
```

Comment:
So it's not important to log a debug message here about a malformed message?

Reply:
Well spotted. I forgot to re-introduce it.